add-ons/updating/tb78/convert-legacy.md

```diff
@@ -0,0 +1,146 @@
+
+Any interaction with Thunderbird must occur through these APIs. Whenever code needs to be added to the background script, you need to make sure to migrate calls to XPCOM or other native Thunderbird features to these APIs.
+
+
+## Experiment-ing with new APIs
+
+While the Thunderbird team plans to add more APIs with upcoming releases, the current set of APIs will not be sufficient to port most add-ons. To work around this limitation, add-ons can introduce their own, additional APIs as so-called [*experiments*](https://thunderbird-webextensions.readthedocs.io/en/latest/how-to/experiments.html). **Any feature that was available in previous versions of Thunderbird remains available in Thunderbird 78 when using experiments.**
+
+Experiments consist of three parts:
+1. a [*schema*](https://firefox-source-docs.mozilla.org/toolkit/components/extensions/webextensions/schema.html) describes the API that can be accessed by the WebExtension part of the add-on. The schema may define functions that can be called by the add-in, events the add-in can register listeners for.
+2. a *parent* implementation implements the API in Thunderbird's main process. All features that were available to a bootstrapped legacy extension can be used here.
+3. a *child* implementation implements the API in the content process(es). In most cases, it will get generated automatically to call the parent implementation – but manual adjustments may be necessary if function parameters cannot be transferred to a different process (for example, if an API function accepts a function as argument). If the API does not need to interact with anything running in the main thread, it is also possible to implement an API completely within the child implementation.
+
+   To return values not supported by the [structured clone algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm), you may need to clone them into the unprivileged scope of the WebExtension. Usually, that boils down to invoking [`Components.utils.cloneInto()`](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Language_Bindings/Components.utils.cloneInto) or a related function with `context.cloneScope` as target scope. Alternatively, you can directly return unprivileged objects created using constructors within `context.cloneScope`.
+
```

Comment:
You lost me here, do we want that much deep stuff in the update guide?

Reply:
I don't know what level of detail is needed. If this level of detail is not needed, I agree that the whole section can be removed (other than the introduction explaining that there are experiments and linking to the Thunderbird WebExtensions page).

For my most complex add-on (GeneralSync's  integration add-on), these details seem to be essential. I'd argue that they are only relevant if you need to do 'complex' interactions between WebExtension and Experiment code, though.

Example: if you want to write an API that returns an object with async methods to the WebExtension, the naive `return {stuff};` approach does not work.

The Mozilla docs do cover these things, but the information is spread out in multiple places, as they approach some topics from the web side of things (privilege boundaries in JS) and some from the add-on side (experiment structure). So having a short overview on what you _might_ need seems helpful to me. You can argue that addressing this elsewhere would be a better idea than handling it in the guide, though...

Reply:
We will push this in a second "advanced" document. What you have written so far should not be lost. But it is too much for the initial update description.

Reply:
Feel free to do so. This should resolve your comment from the thread above, as you can take the generic parts over there as well (or trash them^^).